akka-actor/src/main/scala/akka/actor/ActorCell.scala

```diff
@@ -727,15 +727,25 @@ private[akka] class ActorCell(
       throw new IllegalArgumentException("ActorCell has no props field")
   }
 
-  final protected def clearActorFields(actorInstance: Actor, recreate: Boolean): Unit = {
-    setActorFields(actorInstance, context = null, self = if (recreate) self else system.deadLetters)
+  final protected def clearActorFields(
+      actorInstance: Actor,
+      recreate: Boolean,
+      context: Option[ActorContext] = Some(null)): Unit = {
+    setActorFields(actorInstance, context = context, self = if (recreate) self else system.deadLetters)
     currentMessage = null
     behaviorStack = emptyBehaviorStack
   }
 
-  final protected def setActorFields(actorInstance: Actor, context: ActorContext, self: ActorRef): Unit =
+  final protected def clearActorFieldsOnTerminate(actorInstance: Actor): Unit =
```

Comment:
Do we really need to special case this, can't we just never clear context from `clearActorFields`?
The only other case is in response to restart AFAICS.

Reply:
I gave it try and never clearing the context leads to extra `preStart` calls when an actor recreation fails. Details below.

 https://github.com/akka/akka/blob/c5e9b5918c4188d103a71f70ba6654c968c837cc/akka-actor-tests/src/test/scala/akka/actor/SupervisorSpec.scala#L399 fails with
```
Filter completion error:
Timeout (3 seconds) waiting for ErrorFilter(class akka.actor.PreRestartException,None,Left(Don't wanna!),true)
java.lang.AssertionError: Filter completion error:
Timeout (3 seconds) waiting for ErrorFilter(class akka.actor.PreRestartException,None,Left(Don't wanna!),true)
	at akka.testkit.package$.filterEvents(package.scala:30)
	at akka.testkit.package$.filterEvents(package.scala:39)
	at akka.actor.SupervisorSpec.$anonfun$new$23(SupervisorSpec.scala:429)
```
because `preStart` of the test actor is called twice. 

Looking further, when recreation fails https://github.com/akka/akka/blob/c5e9b5918c4188d103a71f70ba6654c968c837cc/akka-actor/src/main/scala/akka/actor/dungeon/FaultHandling.scala#L257 sets the context to null exactly to prevent extra `preStart`.
preStarts are called from https://github.com/akka/akka/blob/c5e9b5918c4188d103a71f70ba6654c968c837cc/akka-actor/src/main/scala/akka/actor/dungeon/FaultHandling.scala#L68